### PR TITLE
Restaurant full crud

### PR DESCRIPTION
**Pull request by boris-p**

### CODE REVIEW

File: app/controllers/restaurants_controller.rb [hunk 1]
```diff
@@ -12,4 +12,40 @@ def show
       json_response({message: 'object not found'}, 404)
     end
   end
+
+  def create
+    @restaurant = Restaurant.new(restaurant_params)
+    if @restaurant.save
+      json_response({message: 'object created succussfully'}, 201)
+    else
+      json_response({message: "was not able to save. Error is #{@restaurant.errors.full_messages}"}, 400)
```
Comment (ktamir) [inline app/controllers/restaurants_controller.rb]: 👍 for error handling
However, you can probably extract much of the error handling (especially the 404 ones) to a concern - we do it in our services, and there's also an example on: https://scotch.io/tutorials/build-a-restful-json-api-with-rails-5-part-one

File: app/controllers/restaurants_controller.rb [hunk 2]
```diff
@@ -12,4 +12,40 @@ def show
...
+    end
+  end
+
+  def update
+    begin
+      @restaurant = Restaurant.find(params[:id])
```
Comment (ktamir) [inline app/controllers/restaurants_controller.rb]: This line repeats in show, update, destroy
You can declare `before_action :set_restaurant, only: %i[show update destroy]` and add a `set_restaurant` method with `Restaurant.find(....)`

File: spec/requests/restaurants_spec.rb [hunk 3]
```diff
@@ -4,9 +4,10 @@
   let!(:restaurants) { create_list(:restaurant, 5) }
   let(:rest_id) { restaurants.first.id }
 
-  # Test suite for GET /restaurants
+  #----------------------------------------------------
+  # index
```
Comment (ktamir) [inline spec/requests/restaurants_spec.rb]: IMO these comments are redundant, the describe methods define the scope pretty well

File: spec/requests/restaurants_spec.rb [hunk 4]
```diff
@@ -33,4 +37,59 @@
       end
     end
   end
+  #----------------------------------------------------
+  # create
+  #----------------------------------------------------
+  describe "POST /restaurants" do
+    context 'when the restaurant fields are valid' do
+      before { post "/restaurants", params: {name: 'new rest name', url: "http://awesomerest.org",
```
Comment (ktamir) [inline spec/requests/restaurants_spec.rb]: Please use factory bot to create the object

File: spec/requests/restaurants_spec.rb [hunk 5]
```diff
@@ -33,4 +37,59 @@
...
+      before { post "/restaurants", params: {bad_param: 'really bad params'} }
+      it 'returns status code 400' do
+        expect(response).to have_http_status(400)
+      end
+    end
+  end
+  #----------------------------------------------------
+  # update
+  #----------------------------------------------------
+  describe "PATCH /restaurants/:id" do
```
Comment (ktamir) [inline spec/requests/restaurants_spec.rb]: Is PUT supported too in your app? Which one are you willing to use in the frontend?

File: spec/requests/restaurants_spec.rb [hunk 6]
```diff
@@ -33,4 +37,59 @@
...
+  # destroy
+  #----------------------------------------------------
+  describe "DELETE /restaurants/:id" do
+    before { delete "/restaurants/#{rest_id}" }
+
+    context 'when trying to delete a restaurant that exists' do
+      it 'return status code 204' do
+        expect(response).to have_http_status(204)
+      end
+    end
```
Comment (ktamir) [inline spec/requests/restaurants_spec.rb]: Minor: please leave a blank line between contexts (I wonder if rubocop enforces it)

File: app/controllers/restaurants_controller.rb (hunk 2 above)
Comment (boris-p) [reply app/controllers/restaurants_controller.rb]: from what I understand it's a pretty common practice?
it seems a bit counterintuitive to me as the behaviour for the function is scattered across more places. i.e - there's a variable that the function can use, but to know where it's coming from we need to go to another place and see if there's a hook, and then to that hook, and only then we know what it is.
Interested in your thoughts on that

File: spec/requests/restaurants_spec.rb (hunk 5 above)
Comment (boris-p) [reply spec/requests/restaurants_spec.rb]: both are supported but I meant to use patch.
I'm not sure that I got your point here, does it relate only to the tests ?

**Pull request merged**